### ISSUE TITLE
Locking After signing in 

### ISSUE THREAD

**Issue by FatalIgnorance** [dechamps/laplock#3]
Hi, 

If I start my laptop while it's on a docking station when I login laplock will start then lock the device, is there a way to prevent that ? If I login when the laptop is on the docking station it should not lock only once I open and close it again. 

also thank you for all your work on this, it is very much appreciated!!!

**Comment by dechamps** [comment]
Honestly I have no idea. I assume that by docking station you mean you have multiple monitors. That might confuse laplock as it's looking at the power state of the monitor as well. (I never tried to use laplock with more than one monitor so I don't know how it behaves in that case.)

In any case, I'm afraid I can't be of much help - it's been 4 years since I touched this code and I'm not even using it anymore because my current laptop has a firmware bug that prevents laplock from getting the lid state.

**Comment by FatalIgnorance** [comment]
hmm well I see there is a newer release v0.2 maybe that one would work better? Not sure what exactly changed from the different versions. 

Also would it be possible for me to remove the screen monitoring and just keep the lid closing monitor ? How difficult would it be for me to pull that off? Should just be deleting bits of the code right ?

**Comment by FatalIgnorance** [comment]
Thank you for your input, it's ok if you cannot help with the issue I completely understand. 

I have narrowed it down however. It seems when laplock is started if the laptops lid is already closed as it would be when the laptop is plugged into external peripherals via a docking station. Laplock will lock the computer right away, now since it's stored in the startup folder when a user logs into the laptop while it's lid is closed and the statup folder items are loaded, laplock locks the pc. I find this to be a bit strange as I thought laplock listens for the lid state change event rather than simply checking if the lid is closed but I'm no programmer so I don't really know. 

I think one solution would be to run the program as a service instead so that it starts prior to logging in, however it would be nice to have the program work as intended in this circumstance. Guess it's time for me to learn how to program!

**Comment by dechamps** [comment]
If I remember what I did back then correctly, the fact that laplock will lock on startup if the lid is closed is not a bug - it's a feature. The goal is to prevent subtle race conditions where, if you happen to close the lid just before laplock starts, you might end up with an unlocked laptop without realizing it.

Starting laplock before login is an interesting idea and would likely solve this issue, but it's more complicated than it sounds, because a service is not linked to an interactive desktop and therefore can't call [LockWorkStation()](https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-lockworkstation) directly.

**Comment by FatalIgnorance** [comment]
Hmmm, can you think of an easy way to prevent this? As in it should only lock if the lid state changes to closed not already is? 


Any help would be very much appreciated but again I do not expect anything. If you are unable to assist I can try to decipher the code myself.

**Comment by dechamps** [comment]
Looking at the code, what I suspect is happening is Windows fires a `GUID_LIDSWITCH_STATE_CHANGE` as soon as `RegisterPowerSettingNotification()` is called if the lid is already closed. (In other words this behaviour emerges naturally - there is no explicit check for lid state at startup.) A quick & dirty fix could be to ignore these events if they happen, say, less than 1 second after registration.

**Comment by FatalIgnorance** [comment]
Hmmm ok I'll look at trying to delay the code past registration. I think it might be easier to delay the code entirely than to figure out how to ignore the events for a short period of time since I am not a programmer :) 


Thanks for your assistance!

**Comment by dechamps** [comment]
That's not going to work - the event will be buffered and it will eventually fire no matter how long you wait. Hence my suggestion to ignore the event if it fires very early after laplock startup.

**Comment by FatalIgnorance** [comment]
hmm got it, ok I have a jumping off point. 


wish me luck!

**Comment by FatalIgnorance** [comment]
Hi!

I was able to fix the issue myself after a finally understanding how things worked but I'm not sure if it's the best solution and or if there could be any issues associated with my fix this is what I did:

I added the following to the very start of the code directly under namespace

```
auto start_time = std::chrono::high_resolution_clock::now();
auto t5000ms = std::chrono::milliseconds(5000);
auto start_time5 = start_time + t5000ms;

```
Then I added the following right under Static LRESULT CALLBACK windowProcedure...

```
auto current_time = std::chrono::high_resolution_clock::now();

	if (start_time5 > current_time) {
		LogLine() << "Initial startup wait period not ended";
		return DefWindowProc(hWnd, uMsg, wParam, lParam);
	}
```


I took your advice and added a period of time to ignore all events, I chose to go with 5 seconds although it could probably be much less. 

Any feedback at all would be greatly appreciated but not at all expected. Thank you so much for your help I could not have done it without you!

**Comment by dechamps** [comment]
Yep, that should do the trick. There might be cleaner solutions like attempting to detect if another monitor is plugged in but that that would probably require significantly more work.

**Comment by FatalIgnorance** [comment]
Yeah, but I'm no programmer :p 

thanks for all the help! Glad to see you're still super active on here after all this time

**Comment by Devilotx** [comment]
I'm also experiencing this situation, where users are logging in and immediately being logged back out, any chance there is a compiled version with this fix coming or available for me to test?

**Comment by FatalIgnorance** [comment]
See my responses I posted the fix

**Comment by Devilotx** [comment]
Yes, I'm asking if there is a pre-compiled version of the fixed .exe as I'm not a programmer.

**Comment by FatalIgnorance** [comment]
Ah, I'll do that in a bit for you

**Comment by Devilotx** [comment]
Thanks, I eagerly await the file.  I appreciate it

**Comment by FatalIgnorance** [comment]
you can download it here:

https://github.com/FatalIgnorance/laplock/releases/tag/laplock-0.3

Sorry it took so long I've been very busy at work, let me know if it does not work well for you.

**Comment by Devilotx** [comment]
> you can download it here:
> 
> https://github.com/FatalIgnorance/laplock/releases/tag/laplock-0.3
> 
> Sorry it took so long I've been very busy at work, let me know if it does not work well for you.

Sorry to bring this up 3 weeks later, but we are having widespread issues with this updated deployment failing to launch because it's looking for MSVCP140D.DDL  as I've stated, I'm not a programmer of any kind but it appears this new release has a debug flag or requirement? as that file (As far as I can figure out) is a debug version.

**Comment by FatalIgnorance** [comment]
Ah that means the client computers probably don't have C++ runtime
redistributable installed on them.

Alternatively I can compile the program to have the runtime embedded into
the app instead which is what I should have done to begin with. I'll fix
that soon and post an update.

Sorry!

**Comment by Devilotx** [comment]
It's fine! I appreciate the work!

**Comment by FatalIgnorance** [comment]
ok retry it now, I've updated the package in the same location:

https://github.com/FatalIgnorance/laplock/releases/tag/laplock-0.3

I renamed it and to verify the new package is 239KB instead of the old 145KB in size. Let me know if this resolved your issue.

**Comment by tpwo** [comment]
Hi, I see this issue wasn't updated for a few years, but I fixed this problem in a bit different way before reading the issue, so someone might find it useful.

I spent some time in debugger, and it seems there are always a few messages immediately after the program is started (I guess the same happens after logging in when automatic startup kicks in). One of these messages is always about the lid state. So we just ignore it.

I updated the code to ignore the first message about the lid (regardless if the lid is open or closed).

I tested it when docked and undocked with the lid closed and open, and it seems to work fine, i.e. the first closing of the lid always triggered the lock, and I haven't experienced the "initial lock" issue.

Here's my commit: https://github.com/tpwo/laplock/commit/d2bfe5c1965759f063e763e5ea5512fafb2c78a6

And release with the binary (I also removed screen turning off condition, so only closing the lid triggers the lock. If you want both conditions, then you don't need laplock at all, as you can set it in Windows sign-in options, see [here](https://github.com/tpwo/laplock/tree/0.3?tab=readme-ov-file#alternatives)): https://github.com/tpwo/laplock/releases/tag/0.3
